### PR TITLE
[TRTLLM-4987][feat] Support context logits in TRTLLMSampler

### PR DESCRIPTION
**Pull request by dcampora**
# [TRTLLM-4987] Support context logits in TRTLLMSampler

## Description

This PR adds context logits support in TRTLLMSampler:

* An algorithm `handle_context_logits` is created in Python that mimics the C++ couterpart.
* Python bindings custom caster support for all TensorPtr and TensorConstPtr.
* Removed `std::vector` opaque bindings, as no reference parameters are required.
* Adjusted `test_return_logits` to also cover what is supported in TRTLLMSampler currently.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by dcampora** [comment]
CC @amitz-nv

File: tests/unittest/_torch/test_return_logits.py [hunk 1]
```diff
@@ -89,8 +101,11 @@ def test_generate_with_return_logits(gather_context_logits: bool,
     for output in llm.generate(prompts, sampling_params=sampling_params):
         if gather_context_logits:
             assert output.context_logits is not None
-            assert len(prompts[0].split()) + \
-                   1 == output.context_logits.shape[0]
+            # TODO: The intended behaviour is to return all context logits.
+            #       However, the logits received in the sampler do not contain all context logits.
+            #       For now, only the last context token logits are returned.
+            #       When it is fixed, it should be len(prompts[0].split()) + 1.
```
Comment (Funatiq) [inline tests/unittest/_torch/test_return_logits.py:107]: It should be exactly `len(prompt)`.

File: tensorrt_llm/_torch/pyexecutor/handle_context_logits.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            logits: Input logits tensor
+            decoder_buffers: Decoder buffers for storing intermediate results
+
+        Returns:
+            int: Index into logits tensor after processing all requests
+        """
+        logits_index = 0
+
+        # Copy logits into decoderBuffers.logits
+        decoder_buffer_logits = [torch.empty(0)] * len(decoder_buffers.logits)
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/handle_context_logits.py:41]: Shouldn't this be a simple list?

File: tensorrt_llm/_torch/pyexecutor/handle_context_logits.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            #     logits_shape = logits_view.shape
+            #     logits_type = logits_view.dtype
+            #     # decoder_logits = buffer_manager.gpu((req_beam_width, logits_shape[1]), logits_type)
+            #     # tensorrt_llm.runtime.kernels.tile_tensor(decoder_logits, logits_view, req_beam_width, stream)
+            #     decoder_logits = decoder_logits.unsqueeze(0)
+            # else:
+            #     decoder_buffer_logits[seq_slot] = logits_view[:logits_view.shape[0], :1, :logits_view.shape[1]]
+
+        # Needs to be done in bulk for the copy to work
+        decoder_buffers.logits = decoder_buffer_logits
```
Comment (Funatiq) [inline tensorrt_llm/_torch/pyexecutor/handle_context_logits.py:91]: I don't understand this comment.

**Review by Funatiq — APPROVED** [review]

**Review by tongyuantongyu — APPROVED** [review]

**Review by pcastonguay — APPROVED** [review]
